src/nvim/buffer.c

```diff
@@ -1997,19 +2022,22 @@ static char_u *fname_match(regmatch_T *rmp, char_u *name, bool ignore_case)
   return match;
 }
 
-/*
- * find file in buffer list by number
- */
+/// Find a file in the buffer list by buffer number.
 buf_T *buflist_findnr(int nr)
 {
+  char_u key[sizeof(handle_T) * 2 + 1];
+  hashitem_T *hi;
+
   if (nr == 0) {
     nr = curwin->w_alt_fnum;
   }
 
-  FOR_ALL_BUFFERS(buf) {
-    if (buf->b_fnum == nr) {
-      return buf;
-    }
+  snprintf((char *)key, sizeof(key), "%x", nr);
+  hi = hash_find(&buf_hashtab, key);
+
+  if (!HASHITEM_EMPTY(hi)) {
+    return (buf_T *)(hi->hi_key -
+                     ((unsigned)(curbuf->b_key - (char_u *)curbuf)));
```

Comment:
`hi->hi_key - offsetof(buf_t, b_key)` would be clearer, but an actual map would be nicer.